cpp/src/reshape/explode.cu

```diff
@@ -36,119 +36,225 @@
+
+  GatherType empty{0, stream};
+  return build_table(input_table,
+                     explode_column_idx,
+                     sliced_child,
+                     gather_map,
+                     empty,
+                     std::move(pos),
+                     true,
+                     stream,
+                     mr);
+}
+
+std::unique_ptr<table> explode_outer(table_view const& input_table,
+                                     size_type const explode_column_idx,
+                                     bool include_position,
+                                     rmm::cuda_stream_view stream,
+                                     rmm::mr::device_memory_resource* mr)
+{
+  lists_column_view explode_col{input_table.column(explode_column_idx)};
+  auto sliced_child  = explode_col.get_sliced_child(stream);
+  auto counting_iter = thrust::make_counting_iterator(0);
+  auto offsets       = explode_col.offsets_begin();
+
+  // number of nulls or empty lists found so far in the explode column
+  GatherType null_offset(explode_col.size(), stream);
+
+  auto null_iter = thrust::make_transform_iterator(
+    thrust::make_counting_iterator(0),
+    [offsets, offsets_size = explode_col.size() - 1] __device__(int idx) {
+      return (idx > offsets_size || (offsets[idx + 1] != offsets[idx])) ? 0 : 1;
+    });
+  thrust::inclusive_scan(
+    rmm::exec_policy(stream), null_iter, null_iter + sliced_child.size(), null_offset.begin());
+
+  auto null_or_empty_count = null_offset.size() > 0 ? null_offset.back_element(stream) : 0;
+  if (null_or_empty_count == 0) {
+    // performance penalty, run simple explode
+    return include_position ? explode_position(input_table, explode_column_idx, stream, mr)
+                            : explode(input_table, explode_column_idx, stream, mr);
+  }
+
+  auto gather_map_size = sliced_child.size() + null_or_empty_count;
+
+  GatherType gather_map(gather_map_size, stream);
+  GatherType explode_col_gather_map(gather_map_size, stream);
+  PositionType pos(include_position ? gather_map_size : 0, stream, mr);
+
+  // offsets + 1 here to skip the 0th offset, which removes a - 1 operation later.
+  auto offsets_minus_one = thrust::make_transform_iterator(
+    offsets + 1, [offsets] __device__(auto i) { return (i - offsets[0]) - 1; });
+  // Fill in gather map with all the child column's entries
+  thrust::for_each(rmm::exec_policy(stream),
+                   counting_iter,
+                   counting_iter + sliced_child.size(),
+                   [offsets_minus_one,
+                    gather_map             = gather_map.begin(),
+                    explode_col_gather_map = explode_col_gather_map.begin(),
+                    position_array         = pos.begin(),
+                    include_position,
+                    offsets,
+                    null_offset = null_offset.begin(),
+                    null_iter,
+                    offset_size = explode_col.offsets().size() - 1] __device__(auto idx) {
```

Comment:
A for_each over a counting iterator with a huge capture list is a raw loop in disguise. Is there a better algorithm here?

Reply:
Hm, maybe not. You're got potentially 3 outputs, which doesn't really map to an algorithm.

Reply:
Yeah, I wanted it to be less...loop, but there is a lot going on in here and in the name of performance I was trying to keep it dense. I'd love for a magic bullet algorithm here, but I don't see it. If someone else does, I'd love to know.